core/vm/contracts.go

```diff
@@ -358,3 +374,58 @@ func (c *bn256Pairing) Run(input []byte) ([]byte, error) {
 	}
 	return false32Byte, nil
 }
+
+type blake2F struct{}
+
+func (c *blake2F) RequiredGas(input []byte) uint64 {
+	if len(input) != blake2FInputLength {
+		// Input is malformed, we can't read the number of rounds.
+		// Precompile can't be executed so we set its price to 0.
+		return 0
+	}
+
+	rounds := binary.BigEndian.Uint32(input[0:4])
+	return uint64(rounds)
+}
+
+const blake2FInputLength = 213
+
+var errBlake2FIncorrectInputLength = errors.New(
+	"input length for Blake2 F precompile should be exactly 213 bytes",
+)
+
+func (c *blake2F) Run(input []byte) ([]byte, error) {
+	if len(input) != blake2FInputLength {
+		return nil, errBlake2FIncorrectInputLength
+	}
+
+	rounds := binary.BigEndian.Uint32(input[0:4])
+
+	var h [8]uint64
+	for i := 0; i < 8; i++ {
+		offset := 4 + i*8
+		h[i] = binary.BigEndian.Uint64(input[offset : offset+8])
+	}
+
+	var m [16]uint64
+	for i := 0; i < 16; i++ {
+		offset := 68 + i*8
+		m[i] = binary.BigEndian.Uint64(input[offset : offset+8])
+	}
+
+	var t [2]uint64
+	t[0] = binary.BigEndian.Uint64(input[196:204])
+	t[1] = binary.BigEndian.Uint64(input[204:212])
+
+	f := (input[212] != 0)
+
+	compression.F(&h, m, t, f, rounds)
+
+	var output [64]byte
+	for i := 0; i < 8; i++ {
+		offset := i * 8
+		binary.BigEndian.PutUint64(output[offset:offset+8], h[i])
+	}
```

Comment:
This code may be using the wrong byte order, as discussed starting at https://github.com/ethereum/EIPs/pull/2129#discussion_r300456654 .